src/swarm/neo/util/Batch.d

```diff
@@ -0,0 +1,615 @@
+    /***************************************************************************
+
+        Adds a single record field to the batch.
+
+        Params:
+            Field = type of field
+            field = field to add to the batch
+
+    ***************************************************************************/
+
+    private void addField ( Field ) ( Field field )
+    {
+        static if ( isDynamicArrayType!(Field) )
+        {
+            auto len = field.length;
+            this.addBytes((cast(void*)&len)[0 .. size_t.sizeof]);
+            this.addBytes(field[]);
+        }
+        else
+            this.addBytes((cast(void*)&field)[0 .. Field.sizeof]);
+    }
+
+    /***************************************************************************
+
+        Adds the specified bytes to the batch.
+
+        Params:
+            data = slice of data to add to the batch
+
+    ***************************************************************************/
+
+    private void addBytes ( in void[] data )
+    {
+        (*this.buffer) ~= data;
+    }
+}
+
+/*******************************************************************************
+
+    Record batch extractor.
+
+    Params:
+        Record = tuple of types of which a record consists
+
+*******************************************************************************/
+
+public scope class BatchReader ( Record ... )
+{
+    import ocean.core.Enforce;
+
+    static assert(recordFieldsSupported!(Record)());
+
+    /// Unconsumed data remaining in batch. Slice of batch passed to ctor.
+    private Const!(void)[] remaining;
+
+    /***************************************************************************
+
+        Constructor. Sets this instance to extract non-compressed data from the
+        specified batch.
+
+        Params:
+            batch = BatchWriter-created batch to extract records from. The
+                buffer must remain unmodified and accessible for the lifetime of
+                this instance
+
+    ***************************************************************************/
+
+    public this ( in void[] batch )
+    {
+        this.remaining = batch;
+    }
+
+    /***************************************************************************
+
+        Constructor. Sets this instance to extract compressed data from the
+        specified batch.
+
+        Params:
+            lzo = LZO instance to use for decompression
+            batch = BatchWriter-created batch to extract records from
+            decompress_buf = buffer into which the batch is uncompressed. The
+                buffer must remain unmodified and accessible for the lifetime of
+                this instance
+
+    ***************************************************************************/
+
+    public this ( Lzo lzo, in void[] batch, ref void[] decompress_buf )
+    {
+        // Read uncompressed length from first size_t.sizeof bytes.
+        auto uncompressed_len = *(cast(size_t*)(batch.ptr));
```

Comment:
This needs validation since the content of `batch` comes from an external source. (Even the array bounds check is disabled here! :frowning_face:)
For the same reason `Lzo.uncompress` is [unsafe](https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/io/compress/Lzo.d#L113), use [`decompressSafe`](https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/io/compress/Lzo.d#L152) instead.

Reply:
Good catch.

Reply:
Updated.